src/Symfony/Component/Validator/Constraints/NotPwnedValidator.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\RuntimeException;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+/**
+ * Checks if a password has been leaked in a data breach using haveibeenpwned.com's API.
+ * Use a k-anonymity model to protect the password being searched for.
+ *
+ * @see https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class NotPwnedValidator extends ConstraintValidator
+{
+    private const RANGE_API = 'https://api.pwnedpasswords.com/range/%s';
+    private $httpClient;
+
+    /**
+     * @param callable|null $httpClient function(string $url): string (throw an exception in case of error)
+     */
+    public function __construct(?callable $httpClient = null)
+    {
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof NotPwned) {
+            throw new UnexpectedTypeException($constraint, NotPwned::class);
+        }
+
+        if (null !== $value && !is_scalar($value) && !(\is_object($value) && method_exists($value, '__toString'))) {
+            throw new UnexpectedTypeException($value, 'string');
+        }
+
+        $value = (string) $value;
+        if ('' === $value) {
+            return;
+        }
+
+        $httpClient = $this->httpClient;
+
+        $hash = strtoupper(sha1($value));
+        $hashPrefix = substr($hash, 0, 5);
+        $url = sprintf(self::RANGE_API, $hashPrefix);
+
+        $result = $httpClient ? $httpClient($url) : @file_get_contents($url);
+        if (false === $result) {
+            throw new RuntimeException('Problem contacting the Have I been Pwned API.');
```

Comment:
I am not sure that this is a good idea. It would mean ending up with a server error which provides no valuable feedback to the user. Maybe we should rather allow the developer to decide whether the value would be taken as invalid in such a case or not.

Reply:
It's very unlikely that it happens (GitHub and 1password rely on this service for instance). This message will not be displayed to the end user (it will just see a 500 error). However it will help guessing what the problem is when looking to logs.

I don't think it is worth implementing a fallback for such an edge case.

Reply:
How can you say "it's very unlikely to happen" when you have no idea where the application will run? It may run behind firewall, proxy, over unstable connection etc. This is very dangerous assumption for a generic code (framework that may be used _anywhere_).

Reply:
we might assume users opt-in when using this constraint, but i think being able to control this makes sense (mode = THROW | VIOLATION/ERROR | PASS)

Reply:
> It may run behind firewall, proxy, over unstable connection etc. This is very dangerous assumption for a generic code (framework that may be used anywhere).

In this case it will **always** fail, and it will be detected very quickly. And, in this case, it is better to fail and throw than silently ignoring the problem (it will allow the ops team to fix the config).

Reply:
And disabling a security-related feature because a connection is unstable, or a firewall is misconfigured is a very bad idea.

Reply:
> And disabling a security-related feature because a connection is unstable, or a firewall is misconfigured is a very bad idea.

While I do agree with this, if this is at a point where a new customer has to create an account to order something, you don't want to lose money because just after working hours the connection goes down.

Reply:
And, if someone really want to support this, it can already pass a custom HTTP client that will return an empty array in case of network failure.

Reply:
I don't see why we need to let the user see an internal server error. That's just bad UX. If we agree on the fact that this constraint must pass for a user to continue what they are doing, we can just add a constraint violation here with just a bit more generic message. And if we want to make the developer aware of the failure, we can log an error here too.

Reply:
I'd be in favor of letting the user choose the strategy as suggested by @xabbuh, so we can cover any case (and I don't think it'll make the code way more complex). Default should probably be to ignore IMHO, but I'd be fine with any.
A violation might not really make sense though as it won't become valid until the service is reachable again, there is nothing the user can do about it. Throw or pass might be enough with a custom exception on throw and always logging to error when the service is unreachable.

Reply:
while documenting this validator, we should also tell ppl about this behavior and more importantly its corollary: this should not be the only validation rule in place for passwords. Obvious but still :)

Reply:
In the Email Validator, you can enable the strict mode (https://symfony.com/doc/current/reference/constraints/Email.html)
This could be the solution also for this validator in my opinion.

strict mode will throw an error if service is unreachable and normal mode let the user pass.